### ISSUE TITLE
CIOU loss

### ISSUE THREAD

**Issue by PaulZhangIsing** [miemie2013/Keras-DIOU-YOLOv3#9]
我仿照tensorflow-addons的GIOU写了一个CIOU的
用于VGG16SSD

参考了一下您的代码
有几个问题， 在train.py script 里面， 您为什么在81行， CIOU的计算方式，每个都乘以1.0呢？是为了保持都变成float么
第二个是， 我之前的VGG16SSD用的是smooth L1, 看起来这个loss没问题
但是我一换成DIOU/CIOU， 只要一有距离项目， 整个training很快就爆掉了
因为loss很快就增大了，不知道您有没有遇到过这种情况，
但是单独排除又没有问题

那个clip_by_value是用来防止loss太大爆掉的 我也觉得很奇怪 不知道怎么回事

` def calculate_ciou(self, b1, b2):
        '''
        Arguments:
           target (nD tensor): A TensorFlow tensor of any shape containing the ground truth data.
               In this context, the expected tensor has shape `(batch_size, #boxes, 4)` and
               contains the ground truth bounding box coordinates, where the last dimension
               contains `(xmin, xmax, ymin, ymax)`.
           output (nD tensor): A TensorFlow tensor of identical structure to `y_true` containing
               the predicted data, in this context the predicted bounding box coordinates.
        takes in a list of bounding boxes
        but can work for a single bounding box too
        all the boundary cases such as bounding boxes of size 0 are handled.
        '''

        zero = tf.convert_to_tensor(0.0, b1.dtype)
        x1g, y1g, x2g, y2g = tf.unstack(value=b1, num=4, axis=-1)
        x1, y1, x2, y2 = tf.unstack(value=b2, num=4, axis=-1)

        true_width = tf.maximum(zero, x2g - x1g)
        true_height = tf.maximum(zero, y2g - y1g)
        pred_width = tf.maximum(zero, x2 - x1)
        pred_height = tf.maximum(zero, y2 - y1)
        true_area = true_width * true_height
        pred_area = pred_width * pred_height

        ###iou term###
        intersect_ymin = tf.maximum(y1g, y1)
        intersect_xmin = tf.maximum(x1g, x1)
        intersect_ymax = tf.minimum(y2g, y2)
        intersect_xmax = tf.minimum(x2g, x2)
        intersect_width = tf.maximum(zero, intersect_xmax - intersect_xmin)
        intersect_height = tf.maximum(zero, intersect_ymax - intersect_ymin)
        intersect_area = intersect_width * intersect_height

        union_area = true_area + pred_area - intersect_area
        iou = tf.math.divide_no_nan(intersect_area, union_area)

        ###

        ###distance term###
        x_center = (x2 + x1) / 2
        y_center = (y2 + y1) / 2
        x_center_g = (x1g + x2g) / 2
        y_center_g = (y1g + y2g) / 2
        xc1 = tf.minimum(x1, x1g)
        yc1 = tf.minimum(y1, y1g)
        xc2 = tf.maximum(x2, x2g)
        yc2 = tf.maximum(y2, y2g)
        c = tf.pow((xc2 - xc1), 2) + tf.pow((yc2 - yc1), 2)
        d = tf.pow((x_center - x_center_g), 2) + tf.pow((y_center - y_center_g), 2)
        u = tf.math.divide_no_nan(d, c)
        ###

        ###aspect-ratio term###
        arctan = tf.atan(tf.math.divide_no_nan(true_width, true_height)) - tf.atan(tf.math.divide_no_nan(pred_width, pred_height))
        v = (4 / tf.pow(math.pi, 2)) * tf.pow((tf.atan(tf.math.divide_no_nan(true_width, true_height)) -
                                                                      tf.atan(tf.math.divide_no_nan(pred_width, pred_height))), 2)
        s = 1 - iou
        alpha = tf.math.divide_no_nan(v, s + v)
        w_temp = 2 * pred_width
        # ar = (4 / (tf.pow(math.pi, 2))) * tf.pow(arctan, 2)
        ###
        # ar = tf.clip_by_value(alpha * v, -1.0, 1.0)
        ###calculate ciou###
        ciou = iou - (u + alpha * v)
        ciou = tf.clip_by_value(ciou, 0.0, 1.0)
        return ciou
`
